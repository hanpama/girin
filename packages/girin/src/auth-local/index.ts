import * as argon2 from 'argon2';
import * as jwt from 'jsonwebtoken';

import { Model, field } from '../mongodb';
import { ObjectID } from 'bson';
import AuthLocalModule from './module';
import { globalEnvironment } from '@girin/environment';


function getPasswordSalt() {
  return globalEnvironment.get(AuthLocalModule).PASSWORD_SALT;
}
function getJWTSecretKey() {
  return globalEnvironment.get(AuthLocalModule).JWT_SECRET_KEY;
}

export type BaseUserClass<T extends BaseUser> = typeof BaseUser & {
  new(source: any): T;
}

export class BaseUser extends Model {

  @field()
  protected hashedPassword: string;

  public static async createUserWithPassword(source: { [fieldName: string]: any }, password: string) {

    const hashedPassword = await argon2.hash(password + getPasswordSalt());
    const user = await this.insert({ hashedPassword, ...source });
    return user;
  }

  public authenticate(password: string) {
    return argon2.verify(this.hashedPassword, password + getPasswordSalt());
  }

  public async changePassword(newPassword: string) {
    const hashedPassword = await argon2.hash(newPassword + getPasswordSalt());
    this.hashedPassword = hashedPassword;
    await this.$update({ $set: { hashedPassword } });

    return this;
  }

  public async encodeToken(password: string) {
    const authenticated = await this.authenticate(password);
    if (!authenticated) {
      throw new Error('Authentication Error');
    }
    const { hashedPassword, ...rest } = this.$source;
    return jwt.sign(rest, getJWTSecretKey());
  }

  public static decodeToken<T extends BaseUser>(this: BaseUserClass<T>, token: string): T {
    const result: any = jwt.verify(token, getJWTSecretKey());
    console.log(result);
    const user = new this(result);
    user._id = new ObjectID(result._id);
    return user;
  }
}
