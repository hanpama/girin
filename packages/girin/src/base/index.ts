export * from './scalar';
export * from './Definition';
export * from './MetadataStorage';
export * from './TypeExpression';
export * from './structure';
