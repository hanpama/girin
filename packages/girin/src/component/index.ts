export * from './Component';
export * from './decorators';
