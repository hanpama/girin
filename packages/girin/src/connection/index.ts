export * from './PageInfo';
export * from './Edge';
export * from './Connection';
