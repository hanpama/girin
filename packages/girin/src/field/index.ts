export * from './Field';
export * from './InputField';
// export * from './mount';
