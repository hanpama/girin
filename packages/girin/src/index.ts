export { StringScalar, BooleanScalar, FloatScalar, IntScalar, IDScalar } from './builtins/scalar';
export { List, NonNull } from './type-expression/structure';
export * from './base';
export * from './metadata';
export * from './sdl';
export * from './getGraphQLType';
export * from './field';
export * from './type-expression';
