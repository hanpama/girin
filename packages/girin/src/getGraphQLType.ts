import { MetadataStorage } from "./base/MetadataStorage";
import { DefinitionMetadata } from "./base/DefinitionMetadata";
import { globalMetadataStorage } from "./globalMetadataStorage";


export interface TypeBuilderOptions {
  meta?: MetadataStorage;
}

export function getGraphQLType(targetClass: Function, options?: TypeBuilderOptions): any {

  const typeMetadata = globalMetadataStorage.getDefinitionMetadata(DefinitionMetadata, targetClass);
  if (!typeMetadata) {
    throw new Error(`Given class ${targetClass.name} has no corresponding metadata`);
  }
  return typeMetadata.typeInstance;
}
