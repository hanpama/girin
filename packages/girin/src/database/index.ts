import { loadPredefinedConnectionFields } from './connection/relay';
import { getGlobalMetadataStorage } from 'girin-typelinker';

loadPredefinedConnectionFields(getGlobalMetadataStorage());

export * from './model';
export * from './connection';
