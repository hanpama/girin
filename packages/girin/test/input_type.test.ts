import { GraphQLSchema, graphql, printSchema } from "graphql";

import { gql, getGraphQLType, ObjectType, InputType } from "../src";


@InputType.define(gql`
  input NameInput {
    firstName: String!
    lastName: String!
  }
`)
@ObjectType.define(gql`
  type Name {
    firstName: String!
    lastName: String!
  }
`)
class Name {
  firstName: string;
  lastName: string;

  get fullName() {
    return `${this.firstName} ${this.lastName}`;
  }
}

@InputType.define(gql`
  input PersonInput {
    address: String!
    name: ${Name}
  }
`)
@ObjectType.define(gql`
  type Person {
    address: String!
    name: ${Name}
  }
`)
class Person {
  address: string;
  name: Name;
}


@ObjectType.define(gql`
  type Query {            # resolved to NameInput
    formatFullName(input: ${Name}): String!
                                    # resolved to PersonInput
    personInputInstantiated(person: ${Person}): Boolean!
    personNonNullInputWorks(person: ${Person}!): Boolean!
    personNonNullListInputWorks(people: [${Person}!]): Boolean!


    echoPerson(person: ${Person}): Person! # resolved to Person
  }
`)
class Query {
  static formatFullName(_source: null, args: { input: Name }) {
    return args.input.fullName;
  }
  static personInputInstantiated(_source: null, args: { person: Person }) {
    return (args.person instanceof Person) && (args.person.name instanceof Name);
  }
  static personNonNullInputWorks(_source: null, args: { person: Person }) {
    return (args.person instanceof Person) && (args.person.name instanceof Name);
  }
  static personNonNullListInputWorks(_source: null, args: { people: Person[] }) {
    return args.people.reduce((res, person) => {
      return res && (person instanceof Person) && (person.name instanceof Name);
    }, true);
  }
  static echoPerson(_source: null, args: { person: Person }) {
    return args.person;
  }
}

describe('Input type', () => {
  it('generates schema as expected', async () => {
    const schema = new GraphQLSchema({
      query: getGraphQLType(Query),
    });

    expect(printSchema(schema)).toMatchSnapshot();

    const results = await graphql({ schema, source: `
      query {
        formatFullName(input: { firstName: "Foo", lastName: "Bar" })
        personInputInstantiated(person: {
          address: "A",
          name: {
            firstName: "Foo",
            lastName: "Bar"
          }
        })
        personNonNullInputWorks(person: {
          address: "A",
          name: {
            firstName: "Foo",
            lastName: "Bar"
          }
        })
        personNonNullListInputWorks(people: [
          {
            address: "A",
            name: {
              firstName: "Foo",
              lastName: "Bar"
            }
          },
          {
            address: "B",
            name: {
              firstName: "Foo",
              lastName: "Bar"
            }
          }
        ])

        echoPerson(person: {
          address: "A",
          name: {
            firstName: "Foo",
            lastName: "Bar"
          }
        }) {
          address
          name {
            firstName
            lastName
          }
        }
      }
    ` });

    expect(results).toEqual({ data: {
      formatFullName: 'Foo Bar',
      personInputInstantiated: true,
      personNonNullInputWorks: true,
      personNonNullListInputWorks: true,
      echoPerson: {
        address: "A",
        name: {
          firstName: "Foo",
          lastName: "Bar"
        }
      }
    } });
  });
});
