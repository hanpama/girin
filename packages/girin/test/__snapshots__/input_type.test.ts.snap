// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input type generates schema as expected 1`] = `
"type Name {
  firstName: String!
  lastName: String!
}

input NameInput {
  firstName: String!
  lastName: String!
}

type Person {
  address: String!
  name: Name
}

input PersonInput {
  address: String!
  name: NameInput
}

type Query {
  formatFullName(input: NameInput): String!
  personInputInstantiated(person: PersonInput): Boolean!
  personNonNullInputWorks(person: PersonInput!): Boolean!
  personNonNullListInputWorks(people: [PersonInput!]): Boolean!
  echoPerson(person: PersonInput): Person!
}
"
`;
